File: bin/packages.rs
use std::collections::HashSet;
use std::fs;
use std::io;
use std::path::{Path, PathBuf};
use std::time::Duration;

use serde_json::{Map, Value};

use super::{HarnessId, SyncEnv, err};

mod process;
mod source;
mod validate;

use process::{
    install_inferred_import_packages as install_inferred_import_packages_impl,
    install_package_deps, run_package_build,
};
#[cfg(test)]
use source::{
    clone_attempts_for_tests as clone_attempts_for_tests_impl,
    command_for_tests as command_for_tests_impl,
    github_slug_for_tests as github_slug_for_tests_impl,
};
use source::{
    clone_package, package_cache_dir as package_cache_dir_impl, replace_dir_atomically,
    staging_dir_for,
};
#[cfg(test)]
use validate::validate_package_for_tests as validate_package_for_tests_impl;
use validate::{package_has_build_script, package_is_healthy};

const PACKAGE_SOURCE_FILE: &str = "packages.json";
const PACKAGE_CACHE_SUBDIR: &str = ".local/share/agents/pi-packages";
const RESOURCE_KEYS: [&str; 4] = ["extensions", "skills", "prompts", "themes"];

#[derive(Clone, Debug, PartialEq, Eq)]
pub(super) struct PackageManifest {
    pub(super) packages: Vec<String>,
}

pub(super) fn install_inferred_import_packages(dir: &Path, timeout: Duration) -> bool {
    install_inferred_import_packages_impl(dir, timeout)
}

pub(super) fn package_cache_dir(cache_root: &Path, source: &str) -> PathBuf {
    package_cache_dir_impl(cache_root, source)
}

#[cfg(test)]
pub(super) fn github_slug_for_tests(source: &str) -> Option<String> {
    github_slug_for_tests_impl(source)
}

#[cfg(test)]
pub(super) fn command_for_tests(source: &str, target_dir: &Path) -> Vec<String> {
    command_for_tests_impl(source, target_dir)
}

#[cfg(test)]
pub(super) fn clone_attempts_for_tests(
    source: &str,
    target_dir: &Path,
    gh_available: bool,
    outcomes: &[bool],
) -> (bool, Vec<Vec<String>>) {
    clone_attempts_for_tests_impl(source, target_dir, gh_available, outcomes)
}

#[cfg(test)]
pub(super) fn validate_package_for_tests(dir: &Path) -> Result<bool, String> {
    validate_package_for_tests_impl(dir)
}

pub(super) fn bootstrap_packages(sync_env: &SyncEnv) -> bool {
    let manifest_path = sync_env
        .harness(HarnessId::Pi)
        .map(|harness| {
            harness
                .source_root(&sync_env.tools_home)
                .join(PACKAGE_SOURCE_FILE)
        })
        .unwrap_or_else(|| {
            sync_env
                .tools_home
                .join("pi")
                .join("agent")
                .join(PACKAGE_SOURCE_FILE)
        });
    let manifest = match read_package_manifest(&manifest_path) {
        Ok(manifest) => manifest,
        Err(message) => {
            err(&format!("package bootstrap failed: {message}"));
            return false;
        }
    };

    let runtime_settings_path = sync_env
        .harness(HarnessId::Pi)
        .map(|harness| harness.root().join("settings.json"))
        .unwrap_or_else(|| {
            sync_env
                .tools_home
                .join("pi")
                .join("agent")
                .join("settings.json")
        });
    let cache_root = sync_env.home.join(PACKAGE_CACHE_SUBDIR);

    let mut installed_paths = Vec::new();
    let mut success = true;
    for source in &manifest.packages {
        match ensure_package(source, &cache_root, sync_env.install_timeout) {
            Ok(Some(path)) => installed_paths.push(path),
            Ok(None) => {}
            Err(message) => {
                err(&format!("package bootstrap failed for {source}: {message}"));
                success = false;
            }
        }
    }

    if let Err(message) = patch_runtime_settings(&runtime_settings_path, &installed_paths) {
        err(&format!("package settings patch failed: {message}"));
        success = false;
    }

    success
}

pub(super) fn read_package_manifest(path: &Path) -> Result<PackageManifest, String> {
    let content = match fs::read_to_string(path) {
        Ok(content) => content,
        Err(error) if error.kind() == io::ErrorKind::NotFound => {
            return Ok(PackageManifest {
                packages: Vec::new(),
            });
        }
        Err(error) => return Err(format!("{} ({error})", path.display())),
    };

    let value: Value = serde_json::from_str(&content)
        .map_err(|error| format!("invalid JSON in {}: {error}", path.display()))?;
    let object = value
        .as_object()
        .ok_or_else(|| format!("{} must contain a JSON object", path.display()))?;
    let packages_value = object
        .get("packages")
        .ok_or_else(|| format!("{} missing \"packages\" array", path.display()))?;
    let packages_array = packages_value
        .as_array()
        .ok_or_else(|| format!("{} field \"packages\" must be an array", path.display()))?;

    let mut seen = HashSet::new();
    let mut packages = Vec::new();
    for entry in packages_array {
        let source = entry
            .as_str()
            .ok_or_else(|| format!("{} package entries must be strings", path.display()))?
            .trim();
        if source.is_empty() {
            return Err(format!(
                "{} package entries must not be empty",
                path.display()
            ));
        }
        if seen.insert(source.to_string()) {
            packages.push(source.to_string());
        }
    }

    Ok(PackageManifest { packages })
}

pub(super) fn patch_runtime_settings(path: &Path, package_paths: &[PathBuf]) -> Result<(), String> {
    let current = match fs::read_to_string(path) {
        Ok(content) => content,
        Err(error) if error.kind() == io::ErrorKind::NotFound => "{}".to_string(),
        Err(error) => return Err(format!("read {} ({error})", path.display())),
    };
    let mut value: Value = serde_json::from_str(&current)
        .map_err(|error| format!("parse {} ({error})", path.display()))?;
    if !value.is_object() {
        value = Value::Object(Map::new());
    }
    let object = value
        .as_object_mut()
        .ok_or_else(|| format!("{} must contain a JSON object", path.display()))?;
    object.insert(
        "packages".to_string(),
        Value::Array(
            package_paths
                .iter()
                .map(|path| Value::String(path.to_string_lossy().to_string()))
                .collect(),
        ),
    );

    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent)
            .map_err(|error| format!("create {} ({error})", parent.display()))?;
    }
    let content = serde_json::to_string_pretty(&value)
        .map_err(|error| format!("serialize {} ({error})", path.display()))?;
    fs::write(path, format!("{content}\n"))
        .map_err(|error| format!("write {} ({error})", path.display()))
}

fn ensure_package(
    source: &str,
    cache_root: &Path,
    timeout: Duration,
) -> Result<Option<PathBuf>, String> {
    let final_dir = package_cache_dir(cache_root, source);
    if package_is_healthy(&final_dir)? {
        return Ok(Some(final_dir));
    }

    fs::create_dir_all(cache_root)
        .map_err(|error| format!("create cache root {} ({error})", cache_root.display()))?;
    let staging_dir = staging_dir_for(&final_dir);
    super::rm_entry(&staging_dir)
        .map_err(|error| format!("clear staging dir {} ({error})", staging_dir.display()))?;
    fs::create_dir_all(staging_dir.parent().unwrap_or(cache_root))
        .map_err(|error| format!("create staging parent {} ({error})", staging_dir.display()))?;

    if !clone_package(source, &staging_dir, timeout) {
        let _ = super::rm_entry(&staging_dir);
        return Err("clone failed".to_string());
    }
    if !install_package_deps(&staging_dir, timeout) {
        let _ = super::rm_entry(&staging_dir);
        return Err("dependency install failed".to_string());
    }

    let mut healthy = package_is_healthy(&staging_dir)?;
    if !healthy && package_has_build_script(&staging_dir)? {
        if !run_package_build(&staging_dir, timeout) {
            let _ = super::rm_entry(&staging_dir);
            return Err("build failed".to_string());
        }
        if !install_inferred_import_packages(&staging_dir, timeout) {
            let _ = super::rm_entry(&staging_dir);
            return Err("install inferred packages after build failed".to_string());
        }
        healthy = package_is_healthy(&staging_dir)?;
    }

    if !healthy {
        let _ = super::rm_entry(&staging_dir);
        return Err("package resources failed validation".to_string());
    }

    replace_dir_atomically(&staging_dir, &final_dir)
        .map_err(|error| format!("activate package {} ({error})", final_dir.display()))?;
    Ok(Some(final_dir))
}

File: bin/packages.ts
import * as fs from "node:fs";
import path from "node:path";

import { Effect } from "effect";

import {
  installInferredImportPackages as installInferredImportPackagesImpl,
  installPackageDeps,
  runPackageBuild,
} from "./packages/process.ts";
import {
  cloneAttemptsForTests as cloneAttemptsForTestsImpl,
  clonePackage,
  commandForTests as commandForTestsImpl,
  githubSlugForTests as githubSlugForTestsImpl,
  packageCacheDir as packageCacheDirImpl,
  replaceDirAtomically,
  rmEntry,
  stagingDirFor,
} from "./packages/source.ts";
import {
  packageHasBuildScript,
  packageIsHealthy,
  validatePackageForTests as validatePackageForTestsImpl,
} from "./packages/validate.ts";

const PACKAGE_SOURCE_FILE = "packages.json";
const PACKAGE_CACHE_SUBDIR = ".local/share/agents/pi-packages";

export interface PackageManifest {
  readonly packages: string[];
}

export interface HarnessLike {
  readonly id?: unknown;
  readonly sourceName?: string;
  readonly source_name?: string;
  readonly home?: string;
  readonly runtimeSubdir?: string | null;
  readonly runtime_subdir?: string | null;
  sourceRoot?(toolsHome: string): string;
  root?(): string;
}

export interface SyncEnvLike {
  readonly home: string;
  readonly toolsHome?: string;
  readonly tools_home?: string;
  readonly installTimeout?: number;
  readonly installTimeoutMs?: number;
  readonly install_timeout?: number;
  readonly harnesses?: readonly HarnessLike[];
  harness?(id: unknown): HarnessLike | undefined;
}

function err(message: string): void {
  console.error(`sync: ${message}`);
}

function getToolsHome(syncEnv: SyncEnvLike): string {
  return syncEnv.toolsHome ?? syncEnv.tools_home ?? path.join(syncEnv.home, ".config", "agents", "tools");
}

function getHarnessSourceName(harness: HarnessLike, fallback = "pi"): string {
  return harness.sourceName ?? harness.source_name ?? fallback;
}

function getHarnessRoot(harness: HarnessLike, fallbackHome: string): string {
  if (typeof harness.root === "function") {
    return harness.root();
  }

  const home = harness.home ?? fallbackHome;
  const runtimeSubdir = harness.runtimeSubdir ?? harness.runtime_subdir;
  return runtimeSubdir ? path.join(home, runtimeSubdir) : home;
}

function getHarnessSourceRoot(harness: HarnessLike, toolsHome: string, sourceName: string): string {
  if (typeof harness.sourceRoot === "function") {
    return harness.sourceRoot(toolsHome);
  }

  const runtimeSubdir = harness.runtimeSubdir ?? harness.runtime_subdir;
  const base = path.join(toolsHome, sourceName);
  return runtimeSubdir ? path.join(base, runtimeSubdir) : base;
}

function getPiHarness(syncEnv: SyncEnvLike): HarnessLike | undefined {
  if (typeof syncEnv.harness === "function") {
    const candidates = [syncEnv.harness("Pi"), syncEnv.harness("pi"), syncEnv.harness("PI")];
    for (const candidate of candidates) {
      if (candidate) {
        return candidate;
      }
    }
  }

  for (const harness of syncEnv.harnesses ?? []) {
    if (getHarnessSourceName(harness, "") === "pi") {
      return harness;
    }
    if (String(harness.id ?? "").toLowerCase() === "pi") {
      return harness;
    }
  }
  return undefined;
}

export function packageCacheDir(cacheRoot: string, source: string): string {
  return packageCacheDirImpl(cacheRoot, source);
}

export function githubSlugForTests(source: string): string | null {
  return githubSlugForTestsImpl(source);
}

export function commandForTests(source: string, targetDir: string): string[] {
  return commandForTestsImpl(source, targetDir);
}

export function cloneAttemptsForTests(
  source: string,
  targetDir: string,
  ghAvailable: boolean,
  outcomes: readonly boolean[],
): Promise<[boolean, string[][]]> {
  return cloneAttemptsForTestsImpl(source, targetDir, ghAvailable, outcomes);
}

export function validatePackageForTests(dir: string): boolean {
  return validatePackageForTestsImpl(dir);
}

export function readPackageManifest(filePath: string): PackageManifest {
  let content: string;
  try {
    content = fs.readFileSync(filePath, "utf8");
  } catch (error) {
    if (isNotFound(error)) {
      return { packages: [] };
    }
    throw new Error(`${filePath} (${String(error)})`);
  }

  let value: unknown;
  try {
    value = JSON.parse(content) as unknown;
  } catch (error) {
    throw new Error(`invalid JSON in ${filePath}: ${String(error)}`);
  }

  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new Error(`${filePath} must contain a JSON object`);
  }
  const object = value as Record<string, unknown>;
  const packagesValue = object.packages;
  if (packagesValue === undefined) {
    throw new Error(`${filePath} missing "packages" array`);
  }
  if (!Array.isArray(packagesValue)) {
    throw new Error(`${filePath} field "packages" must be an array`);
  }

  const seen = new Set<string>();
  const packages: string[] = [];
  for (const entry of packagesValue) {
    if (typeof entry !== "string") {
      throw new Error(`${filePath} package entries must be strings`);
    }
    const source = entry.trim();
    if (source.length === 0) {
      throw new Error(`${filePath} package entries must not be empty`);
    }
    if (!seen.has(source)) {
      seen.add(source);
      packages.push(source);
    }
  }

  return { packages };
}

export function patchRuntimeSettings(filePath: string, packagePaths: readonly string[]): void {
  let current = "{}";
  try {
    current = fs.readFileSync(filePath, "utf8");
  } catch (error) {
    if (!isNotFound(error)) {
      throw new Error(`read ${filePath} (${String(error)})`);
    }
  }

  let value: unknown;
  try {
    value = JSON.parse(current) as unknown;
  } catch (error) {
    throw new Error(`parse ${filePath} (${String(error)})`);
  }

  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    value = {};
  }

  (value as Record<string, unknown>).packages = packagePaths.map((packagePath) => packagePath.toString());

  const parent = path.dirname(filePath);
  if (parent.length > 0) {
    fs.mkdirSync(parent, { recursive: true });
  }

  try {
    fs.writeFileSync(filePath, `${JSON.stringify(value, null, 2)}\n`);
  } catch (error) {
    throw new Error(`write ${filePath} (${String(error)})`);
  }
}

async function ensurePackage(source: string, cacheRoot: string, timeoutMs: number): Promise<string> {
  const finalDir = packageCacheDir(cacheRoot, source);
  if (packageIsHealthy(finalDir)) {
    return finalDir;
  }

  fs.mkdirSync(cacheRoot, { recursive: true });

  const stagingDir = stagingDirFor(finalDir);
  rmEntry(stagingDir);
  fs.mkdirSync(path.dirname(stagingDir), { recursive: true });

  try {
    if (!(await clonePackage(source, stagingDir, timeoutMs))) {
      throw new Error("clone failed");
    }
    if (!(await Effect.runPromise(installPackageDeps(stagingDir, timeoutMs)))) {
      throw new Error("dependency install failed");
    }

    let healthy = packageIsHealthy(stagingDir);
    if (!healthy && packageHasBuildScript(stagingDir)) {
      if (!(await Effect.runPromise(runPackageBuild(stagingDir, timeoutMs)))) {
        throw new Error("build failed");
      }
      if (!(await Effect.runPromise(installInferredImportPackagesImpl(stagingDir, timeoutMs)))) {
        throw new Error("install inferred packages after build failed");
      }
      healthy = packageIsHealthy(stagingDir);
    }

    if (!healthy) {
      throw new Error("package resources failed validation");
    }

    await replaceDirAtomically(stagingDir, finalDir);
    return finalDir;
  } catch (error) {
    rmEntry(stagingDir);
    throw error;
  }
}

export async function bootstrapPackages(syncEnv: SyncEnvLike): Promise<boolean> {
  const toolsHome = getToolsHome(syncEnv);
  const piHarness = getPiHarness(syncEnv);
  const manifestPath = piHarness
    ? path.join(getHarnessSourceRoot(piHarness, toolsHome, "pi"), PACKAGE_SOURCE_FILE)
    : path.join(toolsHome, "pi", "agent", PACKAGE_SOURCE_FILE);
  const runtimeSettingsPath = piHarness
    ? path.join(getHarnessRoot(piHarness, path.join(syncEnv.home, ".pi", "agent")), "settings.json")
    : path.join(syncEnv.home, ".pi", "agent", "settings.json");
  const cacheRoot = path.join(syncEnv.home, PACKAGE_CACHE_SUBDIR);

  let manifest: PackageManifest;
  try {
    manifest = readPackageManifest(manifestPath);
  } catch (error) {
    err(`package bootstrap failed: ${String(error instanceof Error ? error.message : error)}`);
    return false;
  }

  const installedPaths: string[] = [];
  let success = true;
  for (const source of manifest.packages) {
    try {
      installedPaths.push(
        await ensurePackage(
          source,
          cacheRoot,
          syncEnv.installTimeoutMs ?? syncEnv.installTimeout ?? syncEnv.install_timeout ?? 120_000,
        ),
      );
    } catch (error) {
      err(`package bootstrap failed for ${source}: ${String(error instanceof Error ? error.message : error)}`);
      success = false;
    }
  }

  try {
    patchRuntimeSettings(runtimeSettingsPath, installedPaths);
  } catch (error) {
    err(`package settings patch failed: ${String(error instanceof Error ? error.message : error)}`);
    success = false;
  }

  return success;
}

export async function installInferredImportPackages(
  dir: string,
  timeoutMs: number,
): Promise<boolean> {
  return Effect.runPromise(installInferredImportPackagesImpl(dir, timeoutMs));
}

export const bootstrapPackagesEffect = (syncEnv: SyncEnvLike) =>
  Effect.tryPromise({
    try: () => bootstrapPackages(syncEnv),
    catch: (error) => error as Error,
  });

export { packageHasBuildScript, packageIsHealthy };

function isNotFound(error: unknown): boolean {
  return (
    typeof error === "object" &&
    error !== null &&
    "code" in error &&
    (error as { code?: unknown }).code === "ENOENT"
  );
}
